File: services/nuldoc/nuldoc-src/generators/atom.ts
import { Config } from "../config.ts";
import { Page } from "../page.ts";
import { PostPage } from "../generators/post.ts";
import { SlidePage } from "../generators/slide.ts";
import { dateToRfc3339String } from "../revision.ts";
import AtomPage from "../pages/AtomPage.ts";

export type Feed = {
  author: string;
  icon: string;
  id: string;
  linkToSelf: string;
  linkToAlternate: string;
  title: string;
  updated: string;
  entries: Entry[];
};

export type Entry = {
  id: string;
  linkToAlternate: string;
  published: string;
  summary: string;
  title: string;
  updated: string;
};

const BASE_NAME = "atom.xml";

export function generateFeedPageFromEntries(
  alternateLink: string,
  feedSlug: string,
  feedTitle: string,
  entries: Array<PostPage | SlidePage>,
  site: "default" | "blog" | "slides",
  config: Config,
): Page {
  const entries_: Entry[] = [];
  for (const entry of entries) {
    entries_.push({
      id: `urn:uuid:${entry.uuid}`,
      linkToAlternate: `https://${
        "event" in entry ? config.sites.slides.fqdn : config.sites.blog.fqdn
      }${entry.href}`,
      title: entry.title,
      summary: entry.description,
      published: dateToRfc3339String(entry.published),
      updated: dateToRfc3339String(entry.updated),
    });
  }
  // Sort by published date in ascending order.
  entries_.sort((a, b) => {
    if (a.published < b.published) {
      return 1;
    } else if (a.published > b.published) {
      return -1;
    }
    return 0;
  });
  const feedPath = `${alternateLink}${BASE_NAME}`;
  const feed: Feed = {
    author: config.site.author,
    icon: `https://${config.sites[site].fqdn}/favicon.svg`,
    id: `tag:${
      config.sites[site].fqdn
    },${config.site.copyrightYear}:${feedSlug}`,
    linkToSelf: `https://${config.sites[site].fqdn}${feedPath}`,
    linkToAlternate: `https://${config.sites[site].fqdn}${alternateLink}`,
    title: feedTitle,
    updated: entries_.reduce(
      (latest, entry) => entry.updated > latest ? entry.updated : latest,
      entries_[0].updated,
    ),
    entries: entries_,
  };

  return {
    root: AtomPage({ feed: feed }),
    renderer: "xml",
    site,
    destFilePath: feedPath,
    href: feedPath,
  };
}

File: services/nuldoc/lib/nuldoc/generators/atom.rb
module Nuldoc
  Feed = Data.define(:author, :icon, :id, :link_to_self, :link_to_alternate, :title, :updated, :entries)

  FeedEntry = Data.define(:id, :link_to_alternate, :published, :summary, :title, :updated)

  module Generators
    class Atom
      BASE_NAME = 'atom.xml'.freeze

      def initialize(alternate_link, feed_slug, feed_title, entries, site, config)
        @alternate_link = alternate_link
        @feed_slug = feed_slug
        @feed_title = feed_title
        @entries = entries
        @site = site
        @config = config
      end

      def generate
        feed_entries = @entries.map do |entry|
          fqdn = entry.respond_to?(:event) ? @config.sites.slides.fqdn : @config.sites.blog.fqdn
          FeedEntry.new(
            id: "urn:uuid:#{entry.uuid}",
            link_to_alternate: "https://#{fqdn}#{entry.href}",
            title: entry.title,
            summary: entry.description,
            published: Revision.date_to_rfc3339_string(entry.published),
            updated: Revision.date_to_rfc3339_string(entry.updated)
          )
        end

        feed_entries.sort! { |a, b| [b.published, a.link_to_alternate] <=> [a.published, b.link_to_alternate] }

        site_entry = @config.site_entry(@site)
        feed_path = "#{@alternate_link}#{BASE_NAME}"

        feed = Feed.new(
          author: @config.site.author,
          icon: "https://#{site_entry.fqdn}/favicon.svg",
          id: "tag:#{site_entry.fqdn},#{@config.site.copyright_year}:#{@feed_slug}",
          link_to_self: "https://#{site_entry.fqdn}#{feed_path}",
          link_to_alternate: "https://#{site_entry.fqdn}#{@alternate_link}",
          title: @feed_title,
          updated: feed_entries.map(&:updated).max || feed_entries.first&.updated || '',
          entries: feed_entries
        )

        Page.new(
          root: Pages::AtomPage.new(feed: feed).render,
          renderer: :xml,
          site: @site,
          dest_file_path: feed_path,
          href: feed_path
        )
      end
    end
  end
end
